### PR TITLE
fix(contract preaudit updates): make changes, fixes, upgrade as requested by contracts team

### PR DESCRIPTION
**Pull request by cmwhited**
# Description

Updates to the `Subscriptions.sol` contract based off `Findings` [here](https://www.notion.so/edgeandnode/GSP-Review-subscriptions-contract-be1e77ad238e4b11a9678d2f3aef801b)

### CODE REVIEW

File: contract/contracts/Subscriptions.sol [hunk 1]
```diff
@@ -202,10 +203,10 @@ contract Subscriptions {
         address user = msg.sender;
         require(user != address(0), 'user is null');
 
-        Subscription storage sub = _subscriptions[user];
+        Subscription memory sub = _subscriptions[user];
```
Comment (tmigone) [inline contract/contracts/Subscriptions.sol:206]: Why was this changed? I think using memory here is actually more expensive because you need to copy the data from storage over to memory and we are only using it for read access.

File: contract/contracts/Subscriptions.sol [hunk 2]
```diff
@@ -219,14 +220,17 @@ contract Subscriptions {
...
+
         emit Unsubscribe(user);
     }
 
     // Extend a user's subscription.
     function extend(address user, uint64 end) public {
         require(user != address(0), 'user is null');
         uint64 currentBlock = uint64(block.number);
-        Subscription storage sub = _subscriptions[user];
+        Subscription memory sub = _subscriptions[user];
```
Comment (tmigone) [inline contract/contracts/Subscriptions.sol:233]: Same as above, I think this is more expensive than using storage.

File: contract/contracts/Subscriptions.sol [hunk 3]
```diff
@@ -236,14 +240,15 @@ contract Subscriptions {
...
-        token.transferFrom(msg.sender, address(this), addition);
-
         setEpochs(sub.start, sub.end, -int128(sub.rate));
         setEpochs(sub.start, end, int128(sub.rate));
 
         _subscriptions[user].end = end;
 
+        uint128 addition = sub.rate * (end - sub.end);
+        bool success = token.transferFrom(msg.sender, address(this), addition);
+        require(success);
```
Comment (tmigone) [inline contract/contracts/Subscriptions.sol:250]: revert message

**Review by tmigone — COMMENTED** [review]
🚀 left some comments!

File: contract/contracts/Subscriptions.sol [hunk 4]
```diff
@@ -99,40 +87,53 @@ contract Subscriptions {
...
-            Prelude.min(int64(currentBlock), int64(sub.end)) - int64(sub.start)
+        uint256 len = uint256(
+            SignedMath.max(
+                0,
+                int256(Math.min(block.number, sub.end)) - int64(sub.start)
+            )
         );
-        return sub.rate * uint128(uint256(len));
+        return sub.rate * uint128(len);
     }
```
[the remark this answers is not in the document]
Comment (tmigone) [reply contract/contracts/Subscriptions.sol:97]: Another think I just thought of, if we make this public then you could also change it so instead of taking a struct as parameter it takes the struct members. 

```
// Instead of this
function locked(Subscription storage sub)

// This
function locked(uint64 start, uint64 end, uint128 rate)
```

The only reason i suggest this is because it makes it easier to interact with the contract via Etherscan for example. So it's just a suggestion, I haven't checked how it changes the rest of the code maybe it's not worth it (or maybe you think it's a silly thing haha)

[the remark this answers is not in the document]
Comment (tmigone) [reply contract/contracts/Subscriptions.sol:97]: If we don't want to make it public then we should remove the function as it's not being used.

File: contract/contracts/Subscriptions.sol [hunk 5]
```diff
@@ -99,40 +87,53 @@ contract Subscriptions {
...
 
     // Unlocked tokens for a subscription are not collectable by the contract
     // owner and can be recovered by the user.
     // Defined as `rate * max(0, end - max(block, start))`
-    function unlocked(Subscription storage sub) private view returns (uint128) {
-        uint64 currentBlock = uint64(block.number);
-        int256 len = Prelude.max(
-            0,
-            int64(sub.end) - Prelude.max(int64(currentBlock), int64(sub.start))
+    function unlocked(Subscription memory sub) private view returns (uint128) {
```
[the remark this answers is not in the document]
Comment (tmigone) [reply contract/contracts/Subscriptions.sol:102]: Same as above if we make it public we can consider changing the input parameters

File: contract/contracts/Subscriptions.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol:97]: I think exposing it as a public function to let a user see how many locked tokens they have makes sense. I wonder if we should lock it down to the user though.

[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: Exposed the function, updated the params to break out of the Subscription object. Also added a separate to find the locked tokens for a user: [5b5324c](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/5b5324c91ce50b9c915371be44c0ce35da3928cc)
@tmigone should this be locked down so only a user can see their locked tokens?

File: contract/contracts/Subscriptions.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: see [7a3d93e](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/7a3d93e3fe0d4304a642b1b50db0b3a9e0ad2088)

File: contract/contracts/Subscriptions.sol [hunk 6]
```diff
@@ -99,40 +87,53 @@ contract Subscriptions {
...
             _collectPerEpoch += epoch.delta;
             total += _collectPerEpoch + epoch.extra;
             delete _epochs[_uncollectedEpoch];
+
+            unchecked { ++_uncollectedEpoch; }
         }
 
-        token.transfer(owner, uint128(total));
+        bool success = token.transfer(owner, uint128(total));
+        require(success);
```
[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: [e858ce6](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/e858ce6ce106dcf2cde7ee5cb9019854cf9fcb66)

File: contract/contracts/Subscriptions.sol [hunk 7]
```diff
@@ -99,40 +87,53 @@ contract Subscriptions {
...
-            0,
-            int64(sub.end) - Prelude.max(int64(currentBlock), int64(sub.start))
+    function unlocked(Subscription memory sub) private view returns (uint128) {
+        uint256 len = uint256(
+            SignedMath.max(
+                0,
+                int256(int64(sub.end)) -
+                    int256(Math.max(block.number, sub.start))
+            )
         );
```
[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: @tmigone it would appear that solidity disagrees with you
Screenshot 2023-02-02 at 8 43 09 AM

File: contract/contracts/Subscriptions.sol [hunk 8]
```diff
@@ -99,40 +87,53 @@ contract Subscriptions {
...
-        uint64 currentBlock = uint64(block.number);
-        int256 len = Prelude.max(
-            0,
-            Prelude.min(int64(currentBlock), int64(sub.end)) - int64(sub.start)
+        uint256 len = uint256(
+            SignedMath.max(
+                0,
+                int256(Math.min(block.number, sub.end)) - int64(sub.start)
+            )
         );
```
[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: Same as above. not allowed

File: contract/contracts/Subscriptions.sol (hunk 3 above)
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: [138dc7a](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/138dc7aeeaffe35334d61049e1d20f215e14394c)

File: contract/contracts/Subscriptions.sol [hunk 9]
```diff
@@ -219,14 +220,17 @@ contract Subscriptions {
             delete _subscriptions[user];
         }
 
+        bool success = token.transfer(user, unlocked(sub));
+        require(success);
```
[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: [138dc7a](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/138dc7aeeaffe35334d61049e1d20f215e14394c)

File: contract/contracts/Subscriptions.sol [hunk 10]
```diff
@@ -175,25 +176,25 @@ contract Subscriptions {
...
         _subscriptions[user] = Subscription({
             start: start,
             end: end,
             rate: rate
         });
         setEpochs(start, end, int128(rate));
 
+        uint128 subTotal = rate * (end - start);
+        bool success = token.transferFrom(msg.sender, address(this), subTotal);
+        require(success);
```
[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: [138dc7a](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/138dc7aeeaffe35334d61049e1d20f215e14394c)

File: contract/contracts/Subscriptions.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: [138dc7a](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/138dc7aeeaffe35334d61049e1d20f215e14394c)

File: contract/contracts/Subscriptions.sol [hunk 11]
```diff
@@ -148,13 +149,13 @@ contract Subscriptions {
                                e1^               e2^
         */
 
-        uint64 e = blockToEpoch(uint64(block.number));
-        uint64 e1 = blockToEpoch(start);
+        uint64 e = uint64(blockToEpoch(block.number));
```
[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: [3cb1fda](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/3cb1fda78afeefbfef976613a7b60d6d0375b3d4)

File: contract/contracts/Subscriptions.sol [hunk 12]
```diff
@@ -219,14 +220,17 @@ contract Subscriptions {
             delete _subscriptions[user];
```
[the remark this answers is not in the document]
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: [ca26b46](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/ca26b46c9180afd5f7fa030699dba53371b37500)

File: contract/contracts/Subscriptions.sol (hunk 2 above)
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: [240afab](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/240afabdb803f9397b372df605cf24d6360fce2b)

File: contract/contracts/Subscriptions.sol (hunk 1 above)
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol]: [240afab](https://github.com/edgeandnode/subscription-payments-contract/pull/9/commits/240afabdb803f9397b372df605cf24d6360fce2b)

File: contract/contracts/Subscriptions.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (tmigone) [reply contract/contracts/Subscriptions.sol]: nice! I dont think locking down is necessary, it's all public data anyways and the user address it's not something that can be easily guessed or associated with anyone.

File: contract/contracts/Subscriptions.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (tmigone) [reply contract/contracts/Subscriptions.sol]: ahh forgot `sub.end` was `uint`.

File: contract/contracts/Subscriptions.sol [hunk 13]
```diff
@@ -95,44 +83,95 @@ contract Subscriptions {
...
-    function locked(Subscription storage sub) private view returns (uint128) {
-        uint64 currentBlock = uint64(block.number);
-        int256 len = Prelude.max(
-            0,
-            Prelude.min(int64(currentBlock), int64(sub.end)) - int64(sub.start)
+    /**
+     * Locked tokens for a subscription are collectable by the contract owner
+     * and cannot be recovered by the user.
+     * Defined as `rate * max(0, min(block, end) - start)`
+     * @param _subStart the start block of the active subscription subscription
```
Comment (Theodus) [inline contract/contracts/Subscriptions.sol:90]:
```suggestion
     * @param _subStart the start block of the active subscription
```

File: contract/contracts/Subscriptions.sol [hunk 14]
```diff
@@ -95,44 +83,95 @@ contract Subscriptions {
...
+        Subscription storage sub = _subscriptions[_user];
+        
+        return locked(sub.start, sub.end, sub.rate);
+    }
+
+    /**
+     * Unlocked tokens for a subscription are not collectable by the contract
+     * owner and can be recovered by the user.
+     * Defined as `rate * max(0, end - max(block, start))`
+     * @param _subStart the start block of the active subscription subscription
```
Comment (Theodus) [inline contract/contracts/Subscriptions.sol:122]:
```suggestion
     * @param _subStart the start block of the active subscription
```

File: contract/contracts/Subscriptions.sol [hunk 15]
```diff
@@ -95,44 +83,95 @@ contract Subscriptions {
...
+     * @param _block block used to calculate the epoch to collect the subset of locked tokens
+     */
+    function collect(uint256 _block) public {
         require(msg.sender == owner, 'must be called by owner');
 
-        uint64 currentEpoch = blockToEpoch(uint64(block.number));
+        uint256 currentEpoch = blockToEpoch(_block);
         int128 total = 0;
-        for (; _uncollectedEpoch < currentEpoch; _uncollectedEpoch++) {
+        for (; _uncollectedEpoch < currentEpoch;) {
```
Comment (Theodus) [inline contract/contracts/Subscriptions.sol:160]:
```suggestion
        while (_uncollectedEpoch < currentEpoch) {
```

**Review by Theodus — COMMENTED** [review]
Great stuff! Just some nits I found while catching myself up on the progress

File: contract/contracts/Subscriptions.sol (hunk 14 above)
Comment (cmwhited) [reply contract/contracts/Subscriptions.sol:122]: 🙃
